packages/back-end/src/services/licenseData.ts

```diff
@@ -66,38 +75,128 @@ export async function getLicenseMetaData() {
+        const team = teamIdToTeamMap[teamId];
+        if (!team) {
+          return;
+        }
+
+        // Global Team Role
+        roles.push(team.role);
+
+        // Team project roles
+        if (team.projectRoles) {
+          roles.push(...team.projectRoles.map((pr) => pr.role));
+        }
+      });
+    }
+  });
+
+  return roles;
+}
+
+export async function getUserCodesForOrg(
+  org: OrganizationInterface
+): Promise<LicenseUserCodes> {
+  const fullMembersSet: Set<string> = new Set([]);
+  const readOnlyMembersSet: Set<string> = new Set([]);
+  let invitesSet: Set<string> = new Set([]);
+
+  let organizations: OrganizationInterface[] = [];
+  let users: { id: string; email: string }[] = [];
+  let teams: TeamInterface[] = [];
+
+  if (IS_CLOUD) {
+    organizations = [org];
     const memberIds = org.members.map((member) => member.id);
-    const memberEmails = (await getUsersByIds(memberIds)).map(
-      (user) => user.email
-    );
-    const inviteEmails = org.invites.map((invite) => invite.email);
-    const membersAndInviteEmails = memberEmails.concat(inviteEmails);
-    userLicenseCodes = membersAndInviteEmails.map((email) => {
-      return md5(email).slice(0, 8);
-    });
+    users = await getUsersByIds(memberIds);
+    teams = await getTeamsForOrganization(org.id);
   } else {
-    // Self-Host logic
-    // get all users and invites codes across all orgs in the db
-    // that are part of at least one organization
-    // there may be multiple orgs in case it is a MULTI_ORG site
-    const emails = await getAllUserEmailsAcrossAllOrgs();
-    const userEmailCodes = await Promise.all(
-      emails.map(async (email) => {
-        return md5(email).slice(0, 8);
-      })
-    );
-
-    const inviteEmails = await getAllInviteEmailsInDb();
-    const inviteEmailCodes: string[] = inviteEmails.map((email) => {
-      return md5(email).slice(0, 8);
-    });
-
-    userLicenseCodes = Array.from(
-      new Set(userEmailCodes.concat(inviteEmailCodes))
-    );
+    // Self-Host, might be multi-org so we have to look across all orgs
+    organizations = await getAllOrgMemberInfoInDb();
+    users = await getUserIdsAndEmailsForAllUsersInDb();
+    teams = await getAllTeamRoleInfoInDb();
+  }
+
+  const userIdsToEmailHash = users.reduce(
+    (acc: { [key: string]: string }, user) => {
+      acc[user.id] = md5(user.email).slice(0, 8);
+      return acc;
+    },
+    {}
+  );
+
+  const teamIdToTeamMap = teams.reduce(
+    (
+      acc: {
+        [key: string]: TeamInterface;
+      },
+      team
+    ) => {
+      acc[team.id] = team;
+      return acc;
+    },
+    {}
+  );
+
+  for (const userId of Object.keys(userIdsToEmailHash)) {
+    const roles = getMemberRoles(organizations, userId, teamIdToTeamMap);
+    const isReadOnly = roles.every(isReadOnlyRole);
+    const emailHash = userIdsToEmailHash[userId];
+
+    if (isReadOnly) {
+      readOnlyMembersSet.add(emailHash);
+    } else {
+      fullMembersSet.add(emailHash);
```

Comment:
We changed how we're getting the list of all user ids for self-hosted.  Before it only included users added to at least 1 org.  Now it includes everyone in Mongo (including orphaned users).

I think if `roles` is an empty array, that means they are orphaned and we should not add them to any of the sets.

Reply:
Whoops.  Good catch.  I confirmed that roles is empty for an orphaned user.